marshal.go

```diff
@@ -295,15 +295,36 @@ func (x *GoSNMP) sendOneRequest(packetOut *SnmpPacket,
 				continue
 			}
 
-			// Detect usmStats report PDUs and go out of this function with all data
-			// (usmStatsNotInTimeWindows [1.3.6.1.6.3.15.1.1.2.0] will be handled by the calling
-			// function, and retransmitted.  All others need to be handled by user code)
-			if result.Version == Version3 && len(result.Variables) == 1 && result.PDUType == Report {
+			// While Report PDU was defined by RFC 1905 as part of SNMPv2, it was never
+			// used until SNMPv3. Report PDU's allow a SNMP engine to tell another SNMP
+			// engine that an error was detected while processing an SNMP message.
+			//
+			// The format for a Report PDU is
+			// -----------------------------------
+			// | 0xA8 | reqid | 0 | 0 | varbinds |
+			// -----------------------------------
+			// where:
+			// - PDU type 0xA8 indicates a Report PDU.
+			// - reqid is either:
+			//    The request identifier of the message that triggered the report
+			//    or zero if the request identifier cannot be extracted.
+			// - The variable bindings will contain a single object identifier and its value
+			if result.Version == Version3 && result.PDUType == Report && len(result.Variables) == 1 {
 				switch result.Variables[0].Name {
-				case ".1.3.6.1.6.3.15.1.1.1.0", ".1.3.6.1.6.3.15.1.1.2.0",
-					".1.3.6.1.6.3.15.1.1.3.0", ".1.3.6.1.6.3.15.1.1.4.0",
-					".1.3.6.1.6.3.15.1.1.5.0", ".1.3.6.1.6.3.15.1.1.6.0":
+				case ".1.3.6.1.6.3.15.1.1.1.0":
+					return nil, fmt.Errorf("error: unknown security level (sec_level)")
+				case ".1.3.6.1.6.3.15.1.1.2.0":
+					// usmStatsNotInTimeWindows will be retransmitted
 					break waitingResponse
+				case ".1.3.6.1.6.3.15.1.1.3.0":
+					return nil, fmt.Errorf("error: unknown username (sec_name)")
+				case ".1.3.6.1.6.3.15.1.1.4.0":
+					// usmStatsUnknownEngineIDs wil be retransmitted
+					break waitingResponse
+				case ".1.3.6.1.6.3.15.1.1.5.0":
+					return nil, fmt.Errorf("error: wrong digest (auth_protocol, auth_password)")
+				case ".1.3.6.1.6.3.15.1.1.6.0":
+					return nil, fmt.Errorf("error: decryption error (priv_protocol, priv_password)")
 				}
 			}
```

Comment:
What about other types of reports that are received? shouldn't those be an error as well?

Reply:
Added a default case to the switch to handle these as well now